File: rust/project/naming.rb
require "hecks/vocabulary"

module RustProjection
  module Projector
    module_function

    SCALAR     = { "String" => "String", "Integer" => "i64", "Float" => "f64" }.freeze
    SCALAR_KIND = { "String" => :string, "Integer" => :int, "Float" => :float }.freeze

    # `Reference<X>` is not a scalar per the IR's own vocabulary, but it
    # behaves like one for codegen purposes: aggregates-and-value-objects.md's
    # "Pointing at another aggregate" section is explicit that a reference
    # "is a bare id — a String — not a nested object." So a reference is
    # represented as a plain `String`, the same as any other id, at the
    # STRUCT-FIELD level. `resolve_references` (Ruby's own existence check —
    # `command_rules/references.rb`, read directly) is now generated too,
    # but at the registry level (`reactions.rb`'s `reference_check`,
    # `registry.rb`'s own emit), which is where Ruby's own version reaches
    # through `@registry.repository(domain, target)` — not here, and not by
    # changing this field's Rust type.
    def reference_type?(type_name) = type_name.to_s.start_with?("Reference<")

    # `X` out of `Reference<X>` — the target aggregate's own bare name, the
    # same string `command_rules/references.rb#referenced_aggregate` reaches
    # via `attribute.type.resolve.name`. `nil` for anything that isn't a
    # reference type at all, same shape as `reference_type?`.
    def reference_target(type_name)
      match = type_name.to_s.match(/\AReference<(.+)>\z/)
      match && match[1]
    end

    def effective_scalar_type(type_name)
      return "String" if reference_type?(type_name)

      type_name if SCALAR.key?(type_name)
    end

    def rust_type(type_name, list:)
      scalar = effective_scalar_type(type_name)
      inner = scalar ? SCALAR.fetch(scalar) : rust_ident(type_name)
      list ? "Vec<#{inner}>" : inner
    end

    def rust_ident(name) = name.to_s.gsub(/[^A-Za-z0-9]/, "")
    def dispatch_fn_name(cmd) = cmd.gsub(/(?<=.)([A-Z])/, '_\1').downcase

    # Two different jobs a field name does in generated code, and they must
    # NOT share one function: as a STRING LITERAL (a `Fielded::field` match
    # arm's key, a payload map key) the plain name is correct Rust — `"type"`
    # IS the string a caller passes to look a field up by name. As an
    # IDENTIFIER (a struct field, `self.X`/`args.X`/`record.X`) a name that
    # collides with a Rust keyword is a syntax error unless raw-escaped —
    # found live: the self-hosted grammar's own `Attribute`/`Argument`
    # aggregates declare a field literally named `type`, because that is
    # what it is.
    #
    # Declared once, as the `RustReservedWord` vocabulary (language/bluebook/
    # vocabulary.bluebook) — the same table bin/project_reserved_names
    # projects into hecks-codegen's `reserved_names.rs`.
    RUST_KEYWORDS = Hecks::Vocabulary.fetch("RustReservedWord")

    # A DOMAIN'S OWN NAME (`File.basename(domain)`, `bin/project_rust`) has
    # to double as THREE things at once: a directory name, a bare Rust
    # module identifier (`pub mod #{name};` — module names get no raw-
    # identifier `r#name` escape hatch the way struct FIELDS do via
    # `rust_ident_field`), and a Cargo `[features]` key. `RUST_KEYWORDS`
    # above rules out the second; this list rules out the third — every
    # key that already appears elsewhere in `rust/Cargo.toml` outside
    # `[features]` (`[package]`'s `name`/`version`/`edition`/etc., `[lib]`/
    # `[[bin]]`'s `name`/`path`) plus `default`, which Cargo itself
    # reserves for the auto-enabled feature set. `bin/project_rust`'s own
    # feature-sync regex used to search the WHOLE Cargo.toml text for
    # `/^name\s*=/` (now scoped to `[features]` only — see its own
    # comment), so a domain named one of these used to silently never
    # get its own feature line while still being set as `default`,
    # producing a Cargo.toml that referenced an undeclared feature.
    # Declared once, as the `CargoReservedName` vocabulary.
    CARGO_RESERVED_DOMAIN_NAMES = Hecks::Vocabulary.fetch("CargoReservedName")

    # TRUE exactly when `name` is safe to use, unescaped, as all three of
    # the above at once. Plain lowercase-identifier shape (no leading
    # digit, no `-`, not empty) rules out landmine class 1 (an outright
    # Rust syntax error); the two reserved-word checks rule out class 2
    # (a legal identifier that collides with something load-bearing).
    def valid_domain_mod_name?(name)
      str = name.to_s
      str.match?(/\A[a-z_][a-z0-9_]*\z/) &&
        !RUST_KEYWORDS.include?(str) &&
        !CARGO_RESERVED_DOMAIN_NAMES.include?(str)
    end

    # BUG#124 — AN AGGREGATE'S OWN NAME has the SAME landmine class 2 as a
    # domain name's `pub mod #{name};` above, at a different site:
    # `domain_generator.rb`'s own per-aggregate file (`#{aggregate[:name].
    # downcase}.rs`) and the `pub mod #{a[:name].downcase};` line it later
    # writes into the domain's own `mod.rs` (both keyed off the DOWNCASED
    # name — an aggregate is conventionally declared PascalCase, e.g.
    # `Crate`, and it's the lowercase form that collides with a keyword).
    # An aggregate name does NOT also have to double as a Cargo feature
    # key (only a DOMAIN's own directory name does — see
    # `CARGO_RESERVED_DOMAIN_NAMES`'s own header), so this checks only
    # the identifier-shape + `RUST_KEYWORDS` half of `valid_domain_mod_
    # name?`'s own two checks, against the downcased form actually used
    # as the module identifier — not the raw, as-declared name.
    #
    # Unescapable in EITHER direction: module names get no `r#name` raw-
    # identifier escape hatch at all (same as a domain name, per `bin/
    # project_rust`'s own comment on `target_mod_name`), and even if they
    # did, `crate`/`self`/`super`/`Self` (`RUST_UNESCAPABLE_KEYWORDS`,
    # below) couldn't use one anyway — raw identifiers are syntactically
    # excluded for exactly those four. Refusing up front, the same way
    # `valid_domain_mod_name?` already does for a domain name, is the
    # only fix that's actually correct for the whole `RUST_KEYWORDS` set,
    # not just the four that could never be escaped either way.
    def valid_aggregate_mod_name?(name)
      str = name.to_s.downcase
      str.match?(/\A[a-z_][a-z0-9_]*\z/) && !RUST_KEYWORDS.include?(str)
    end

    # THE SUBSET OF `RUST_KEYWORDS` A RAW IDENTIFIER (`r#name`) CANNOT
    # RESCUE AT ALL — per the Rust reference's own RAW_IDENTIFIER
    # grammar, `r#crate`/`r#self`/`r#super`/`r#Self` are not merely
    # wrong in some positions, they are not valid raw-identifier SYNTAX,
    # full stop, in any position. `rust_ident_field`, below, escapes
    # every OTHER `RUST_KEYWORDS` entry (`type` -> `r#type`, `fn` ->
    # `r#fn`, ...) correctly — this list is what it must refuse instead
    # of silently emitting broken source for. Not currently reachable
    # from any real corpus field name (verified: no attribute in
    # examples/ or qa/stress_domains/ is named any of these), but the
    # SAME landmine class as BUG#124's aggregate-name collision, at the
    # struct-field site rather than the module-name site — worth closing
    # now that the mechanism is already being extended, rather than
    # leaving a second copy of the same gap for the next generated
    # domain to find live.
    RUST_UNESCAPABLE_KEYWORDS = %w[crate self super Self].freeze

    def rust_field(name) = name.to_s
    def rust_ident_field(name)
      field = rust_field(name)
      if RUST_UNESCAPABLE_KEYWORDS.include?(field)
        raise "rust_ident_field(#{field.inspect}): #{field.inspect} is a Rust keyword that cannot be rescued by a raw " \
              "identifier (r##{field} is not valid Rust syntax for crate/self/super/Self in any position) — rename " \
              "the attribute/field in the bluebook."
      end

      RUST_KEYWORDS.include?(field) ? "r##{field}" : field
    end

    # A closed-set member's value is business text, not a pre-sanitized Rust
    # identifier — splitting on `_`/whitespace alone was enough for every
    # value pizzas/banking ever declared ("small", "large", "open"...), but
    # the self-hosted grammar's own Vocabulary chapter has closed sets whose
    # members are glob patterns ("*.port", "Translations/*.bluebook") —
    # `.capitalize` leaves `*`/`.`/`/` untouched, so the OLD version emitted
    # `Translations/*.bluebook` as a literal enum variant line, and Rust's
    # lexer read the embedded `/*` as an unterminated block comment. Split on
    # ANY run of non-alphanumeric characters instead, so every character that
    # isn't a valid identifier constituent is a word boundary, not preserved
    # text.
    def closed_set_variant(row)
      _, value = row.first
      value.to_s.split(/[^A-Za-z0-9]+/).reject(&:empty?).map(&:capitalize).join
    end

    def screaming_snake(name)
      name.to_s.gsub(/([a-z0-9])([A-Z])/, '\1_\2').upcase
    end

    def scalar_to_value(type_name, rust_expr)
      case type_name
      when "String"  then "Value::Str(#{rust_expr}.clone())"
      when "Integer" then "Value::Int(#{rust_expr})"
      when "Float"   then "Value::Float(#{rust_expr})"
      end
    end

    # TRUE for anything `fielded.rb`'s three "is this attribute a nested
    # value object" sites can safely emit `Field::Nested(&self.x)` for —
    # every ordinary (non-closed-set) VO always could; a single-field
    # CLOSED SET (an enum, `emit_closed_set_enum`) now can too, once it
    # carries its own `Fielded` impl (`emit_closed_set_fielded_impl`,
    # below) answering "value" the same way any other single-field VO's
    # own generated `Fielded` impl already does. FALSE for a MULTI-FIELD
    # closed set (`emit_closed_set_table` — Syntax::Keyword, Argument):
    # genuinely no `Fielded` impl exists or is proven needed for that
    # shape (json_codec.rb's own `emit_closed_set_codec` header: "nothing
    # in either example domain looks one up generically or passes one as
    # a command argument") — `Field::Nested` would fail to compile
    # against it, and nothing in the corpus has ever needed it to.
    #
    # FOUND LIVE, a real, previously-invisible gap: no domain ever
    # declared a `given`/expression referencing a closed-set-typed
    # attribute before lifeadelics' vendored embryonaut_bluebooks/
    # payments (`Payment::Succeed`'s own "the processor matches..."
    # given, over `processor: Processor` and `reported_processor:
    # Processor`, both `one_of:` closed sets) — found live, a reaction-
    # triggered Payment::Succeed refusing with "cannot resolve
    # \"processor\" — no such attribute or argument" even though the
    # aggregate held a real, rehydrated `processor` value the whole time.
    def fielded_capable_nested?(vo)
      !vo[:closed_set] || vo[:attributes].size == 1
    end

    # THE `Fielded` IMPL A SINGLE-FIELD CLOSED SET NEVER GOT —
    # `emit_value_object`'s own single-field-closed-set branch (types.rb)
    # used to return early with JUST the bare enum (`emit_closed_set_
    # enum`), matching neither `emit_fielded_flat`'s shape (it has no
    # `attributes` to iterate, being a tag, not a struct) nor needing one
    # before `fielded_capable_nested?` above started asking. Answers
    # "value" — the SAME single field every other closed-set codec
    # already treats it as (`emit_closed_set_codec`'s own `to_json`:
    # `Json::obj(vec![("value", ...)])`) — with the identical match this
    # enum's own `to_json` already builds, reproduced here rather than
    # shared via a common helper method (lower-risk: purely additive,
    # touches nothing about the enum's own EXISTING to_json/from_json
    # generation or any other caller of it).
    def emit_closed_set_fielded_impl(vo)
      name = rust_ident(vo[:name])
      arms = vo[:members].map do |row|
        _, raw = row.first
        "#{name}::#{closed_set_variant(row)} => #{rust_string_literal(raw.to_s)}.to_string(),"
      end.join(" ")

      "impl crate::kernel::Fielded for #{name} {\n" \
        "    fn field(&self, name: &str) -> Option<crate::kernel::Field<'_>> {\n" \
        "        use crate::kernel::{Field, Value};\n" \
        "        match name {\n" \
        "            \"value\" => Some(Field::Value(Value::Str(match self { #{arms} }))),\n" \
        "            _ => None,\n" \
        "        }\n" \
        "    }\n" \
        "    fn as_scalar(&self) -> Option<crate::kernel::Value> {\n" \
        "        match self.field(\"value\") { Some(crate::kernel::Field::Value(v)) => Some(v), _ => None }\n" \
        "    }\n" \
        "}"
    end

    def literal_rhs(literal)
      case literal
      when String        then "#{rust_string_literal(literal)}.to_string()"
      when Integer, Float then literal.to_s
      when true, false     then literal.to_s
      else raise "unsupported literal mutation source #{literal.inspect} — not one of String/Integer/Float/Boolean"
      end
    end

    # RUST STRING LITERAL ESCAPING (R5) — free-form bluebook text (a
    # `given`/`ensures`/invariant description, a `then_set ... to:`
    # literal, a closed-set member's raw value, the domain's whole IR
    # embedded verbatim as `IR_JSON`) has to become a Rust string literal
    # somewhere in generated code. Ruby's own `String#inspect` LOOKS like
    # it does this job — it already backslash-escapes `"`/`\`/newlines/
    # tabs — but it escapes for RUBY'S OWN benefit, not Rust's:
    # verified live —
    #
    #   "cost #{"#"}{x}".inspect  => "\"cost \\\#{x}\""   (Ruby needs \# so
    #                                the re-read literal doesn't interpolate)
    #   "\x01".inspect            => "\"\\u0001\""        (Ruby's own \uXXXX,
    #                                no braces)
    #
    # Neither `\#` nor a brace-less `\uXXXX` is a legal Rust escape — Rust
    # has no `\#` escape at all, and requires `\u{XXXX}` — so a
    # description containing a literal `#{`/`#@` or a raw control
    # character produces generated Rust that fails to compile (the exact
    # R5 landmine). This escapes only what RUST's own string-literal
    # grammar actually needs — backslash, double-quote, and control
    # characters (`\n`/`\r`/`\t` where they have a short escape, braced
    # `\u{XX}` otherwise) — and leaves everything else, including a
    # literal `#{`/`#@` (meaningless in Rust — it has no string
    # interpolation), untouched.
    def rust_string_literal(str)
      escaped = str.to_s.each_char.map do |ch|
        case ch
        when "\\" then "\\\\"
        when "\"" then "\\\""
        when "\n" then "\\n"
        when "\r" then "\\r"
        when "\t" then "\\t"
        else
          cp = ch.ord
          cp < 0x20 || cp == 0x7F ? format("\\u{%x}", cp) : ch
        end
      end.join
      "\"#{escaped}\""
    end
  end
end

File: rust/codegen/src/naming.rs
//! Port of `rust/project/naming.rb` — read that file's own comments for
//! the full rationale behind each rule; this mirrors its algorithm
//! directly, function for function.

pub fn scalar_rust_type(type_name: &str) -> Option<&'static str> {
    match type_name {
        "String" => Some("String"),
        "Integer" => Some("i64"),
        "Float" => Some("f64"),
        _ => None,
    }
}

pub fn reference_type(type_name: &str) -> bool {
    type_name.starts_with("Reference<")
}

/// `X` out of `Reference<X>` — `None` for anything that isn't a reference
/// type at all.
pub fn reference_target(type_name: &str) -> Option<&str> {
    type_name.strip_prefix("Reference<").and_then(|rest| rest.strip_suffix('>'))
}

pub fn effective_scalar_type(type_name: &str) -> Option<&'static str> {
    if reference_type(type_name) {
        return Some("String");
    }
    match type_name {
        "String" => Some("String"),
        "Integer" => Some("Integer"),
        "Float" => Some("Float"),
        _ => None,
    }
}

pub fn rust_type(type_name: &str, list: bool) -> String {
    let scalar = effective_scalar_type(type_name);
    let inner = match scalar {
        Some(s) => scalar_rust_type(s).unwrap().to_string(),
        None => rust_ident(type_name),
    };
    if list {
        format!("Vec<{inner}>")
    } else {
        inner
    }
}

pub fn rust_ident(name: &str) -> String {
    name.chars().filter(|c| c.is_ascii_alphanumeric()).collect()
}

/// `cmd.gsub(/(?<=.)([A-Z])/, '_\1').downcase` — insert `_` before every
/// uppercase letter that isn't the very first character, then lowercase.
pub fn dispatch_fn_name(cmd: &str) -> String {
    let mut out = String::new();
    for (i, c) in cmd.chars().enumerate() {
        if i > 0 && c.is_ascii_uppercase() {
            out.push('_');
        }
        out.push(c);
    }
    out.to_lowercase()
}

/// Generated from the `RustReservedWord`/`CargoReservedName` vocabularies by
/// `bin/project_reserved_names` — the same tables `naming.rb` reads.
pub use crate::reserved_names::{CARGO_RESERVED_DOMAIN_NAMES, RUST_KEYWORDS};

/// `/\A[a-z_][a-z0-9_]*\z/` — a plain lowercase Rust identifier.
fn plain_lower_identifier(name: &str) -> bool {
    let mut chars = name.chars();
    matches!(chars.next(), Some(c) if c == '_' || c.is_ascii_lowercase())
        && chars.all(|c| c == '_' || c.is_ascii_lowercase() || c.is_ascii_digit())
}

/// Mirrors `legal_aggregate_mod_identifier?` — the identifier-shape half
/// only, against the downcased name.
pub fn legal_aggregate_mod_identifier(name: &str) -> bool {
    plain_lower_identifier(&name.to_lowercase())
}

/// Mirrors `valid_domain_mod_name?` — shape, and neither a Rust keyword nor
/// a reserved Cargo.toml key.
pub fn valid_domain_mod_name(name: &str) -> bool {
    plain_lower_identifier(name) && !RUST_KEYWORDS.contains(&name) && !CARGO_RESERVED_DOMAIN_NAMES.contains(&name)
}

/// Port of `RustProjection::Projector.reserved_name_refusal` (BUG#124) —
/// `None` when every aggregate name and the domain's module name are usable,
/// else the byte-identical refusal the Ruby generator raises. Aggregates are
/// checked (and reported) before the domain name, all at once.
pub fn reserved_name_refusal(source_label: &str, mod_name: &str, aggregate_names: &[&str]) -> Option<String> {
    let refused: Vec<&str> = aggregate_names
        .iter()
        .copied()
        .filter(|name| {
            let module = name.to_lowercase();
            RUST_KEYWORDS.contains(&module.as_str()) || !legal_aggregate_mod_identifier(name)
        })
        .collect();
    if let Some(first) = refused.first() {
        let names = refused.iter().map(|name| format!("{name:?}")).collect::<Vec<_>>().join(", ");
        return Some(format!(
            "{source_label}: aggregate name(s) {names} can't be used as-is — downcased, each becomes a bare Rust \
             module identifier (`pub mod {};`) and a generated file name, and at least one is not a plain identifier \
             or is a Rust keyword (RustReservedWord). Module names get no raw-identifier (r#name) escape hatch — \
             rename the aggregate.",
            first.to_lowercase()
        ));
    }
    if valid_domain_mod_name(mod_name) {
        return None;
    }
    Some(format!(
        "{source_label}: domain module name {mod_name:?} can't be used as-is — it has to double as a Rust module \
         identifier and a Cargo feature name, and this one is either not a plain lowercase identifier, is a Rust \
         keyword (RustReservedWord), or is a reserved Cargo.toml key (CargoReservedName). Rename the domain."
    ))
}

pub fn rust_field(name: &str) -> String {
    name.to_string()
}

pub fn rust_ident_field(name: &str) -> String {
    let field = rust_field(name);
    if RUST_KEYWORDS.contains(&field.as_str()) {
        format!("r#{field}")
    } else {
        field
    }
}

/// A closed-set member's value is business text, not a pre-sanitized Rust
/// identifier — split on any run of non-alphanumeric characters (see
/// naming.rb's own header on why plain `_`/whitespace splitting broke on
/// glob-shaped members like `"*.port"`).
///
/// `Self` — the one capitalized Rust keyword — is renamed by spelling:
/// `SelfType` / `SelfValue` (naming.rb's own `closed_set_variant` header).
pub fn closed_set_variant(value: &str) -> String {
    let variant = value
        .split(|c: char| !c.is_ascii_alphanumeric())
        .filter(|s| !s.is_empty())
        .map(capitalize)
        .collect::<Vec<_>>()
        .join("");
    if variant != "Self" {
        return variant;
    }
    if value.starts_with('S') { "SelfType".to_string() } else { "SelfValue".to_string() }
}

fn capitalize(word: &str) -> String {
    let mut chars = word.chars();
    match chars.next() {
        None => String::new(),
        Some(first) => first.to_uppercase().collect::<String>() + &chars.as_str().to_lowercase(),
    }
}

/// Port of `Hecks::Naming.snake` —
/// `gsub(/([A-Z]+)([A-Z][a-z])/, '\1_\2').gsub(/([a-z\d])([A-Z])/, '\1_\2').downcase`.
pub fn snake(text: &str) -> String {
    let chars: Vec<char> = text.chars().collect();
    let mut first = String::new();
    for (i, &c) in chars.iter().enumerate() {
        let next_upper = chars.get(i + 1).is_some_and(|n| n.is_ascii_uppercase());
        let then_lower = chars.get(i + 2).is_some_and(|n| n.is_ascii_lowercase());
        first.push(c);
        if c.is_ascii_uppercase() && next_upper && then_lower {
            first.push('_');
        }
    }
    let chars: Vec<char> = first.chars().collect();
    let mut second = String::new();
    for (i, &c) in chars.iter().enumerate() {
        second.push(c);
        let next_upper = chars.get(i + 1).is_some_and(|n| n.is_ascii_uppercase());
        if (c.is_ascii_lowercase() || c.is_ascii_digit()) && next_upper {
            second.push('_');
        }
    }
    second.to_lowercase()
}

/// `name.to_s.gsub(/([a-z0-9])([A-Z])/, '\1_\2').upcase`
pub fn screaming_snake(name: &str) -> String {
    let chars: Vec<char> = name.chars().collect();
    let mut out = String::new();
    for (i, &c) in chars.iter().enumerate() {
        if i > 0 {
            let prev = chars[i - 1];
            if (prev.is_ascii_lowercase() || prev.is_ascii_digit()) && c.is_ascii_uppercase() {
                out.push('_');
            }
        }
        out.push(c);
    }
    out.to_uppercase()
}

pub fn scalar_to_value(type_name: &str, rust_expr: &str) -> Option<String> {
    match type_name {
        "String" => Some(format!("Value::Str({rust_expr}.clone())")),
        "Integer" => Some(format!("Value::Int({rust_expr})")),
        "Float" => Some(format!("Value::Float({rust_expr})")),
        _ => None,
    }
}

/// Mirrors `fielded_capable_nested?`: ordinary value objects and
/// single-field closed-set enums implement `Fielded`; multi-field closed-set
/// tables do not.
pub fn fielded_capable_nested(vo: &crate::json::Json) -> bool {
    !vo.get("closed_set").map(crate::json::Json::as_bool).unwrap_or(false)
        || vo.get("attributes").map(crate::json::Json::each).unwrap_or(&[]).len() == 1
}

/// The `Fielded` implementation Ruby emits beside a single-field closed-set
/// enum. Its generic field surface is the same `"value"` object shape used by
/// that enum's existing JSON codec.
pub fn emit_closed_set_fielded_impl(vo: &crate::json::Json) -> String {
    let name = rust_ident(vo.get("name").and_then(crate::json::Json::as_str).unwrap_or(""));
    let arms = vo.get("members").map(crate::json::Json::each).unwrap_or(&[]).iter().filter_map(|row| {
        let pair = row.as_array()?.first()?.as_array()?;
        let raw = pair.get(1)?.to_s();
        Some(format!("{name}::{} => {}.to_string(),", closed_set_variant(&raw), ruby_inspect_string(&raw)))
    }).collect::<Vec<_>>().join(" ");

    format!(
        "impl crate::kernel::Fielded for {name} {{\n    fn field(&self, name: &str) -> Option<crate::kernel::Field<'_>> {{\n        use crate::kernel::{{Field, Value}};\n        match name {{\n            \"value\" => Some(Field::Value(Value::Str(match self {{ {arms} }}))),\n            _ => None,\n        }}\n    }}\n    fn as_scalar(&self) -> Option<crate::kernel::Value> {{\n        match self.field(\"value\") {{ Some(crate::kernel::Field::Value(v)) => Some(v), _ => None }}\n    }}\n}}"
    )
}

/// A literal mutation-source RHS — mirrors `naming.rb#literal_rhs`. Takes
/// the already-parsed `crate::ruby_value::Value`-shaped JSON literal
/// (String/Integer/Float/Bool) this crate reads out of `ir.json`.
pub fn literal_rhs(literal: &crate::json::Json) -> String {
    match literal {
        crate::json::Json::String(s) => format!("{}.to_string()", ruby_inspect_string(s)),
        // `Integer, Float then literal.to_s` — one Ruby case, but `to_s`
        // renders differently per real type (`0.to_s == "0"`, `0.0.to_s ==
        // "0.0"`) — kept as two Rust match arms so a whole-number Float
        // default still renders as a valid `f64` literal (`0.0`, not the
        // integer-typed `0` that would fail to unify with the `x.as_f64()`
        // branch it sits beside in `scalar_from_json_expr`'s own output).
        crate::json::Json::Int(n) => n.to_string(),
        crate::json::Json::Float(_) => literal.to_s(),
        crate::json::Json::Bool(b) => b.to_string(),
        other => panic!("unsupported literal mutation source {other:?} — not one of String/Integer/Float/Boolean"),
    }
}

/// Ruby's `String#inspect` — a double-quoted, backslash/quote-escaped
/// literal. Shared by every codegen site that needs to embed a Ruby-
/// literal-shaped string constant in generated Rust text (matching Rust's
/// own escaping for the common cases both languages agree on).
pub fn ruby_inspect_string(s: &str) -> String {
    let mut out = String::with_capacity(s.len() + 2);
    out.push('"');
    for c in s.chars() {
        match c {
            '"' => out.push_str("\\\""),
            '\\' => out.push_str("\\\\"),
            '\n' => out.push_str("\\n"),
            '\t' => out.push_str("\\t"),
            '\r' => out.push_str("\\r"),
            _ => out.push(c),
        }
    }
    out.push('"');
    out
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn refuses_an_aggregate_whose_module_name_is_a_rust_keyword() {
        let refusal = reserved_name_refusal("shop", "shop", &["Pizza", "Match", "Type"]).expect("refused");
        assert!(refusal.starts_with("shop: aggregate name(s) \"Match\", \"Type\" can't be used as-is"), "{refusal}");
        assert!(refusal.contains("`pub mod match;`"), "{refusal}");
        assert!(refusal.contains("Rust keyword"), "{refusal}");
    }

    #[test]
    fn refuses_an_aggregate_name_that_is_not_a_plain_identifier() {
        assert!(reserved_name_refusal("shop", "shop", &["My-App"]).is_some());
        assert!(reserved_name_refusal("shop", "shop", &[""]).is_some());
    }

    #[test]
    fn refuses_a_domain_module_name_that_is_a_keyword_or_a_cargo_key() {
        for name in ["crate", "type", "package", "default", "version"] {
            let refusal = reserved_name_refusal("label", name, &["Widget"]).expect("refused");
            assert!(refusal.starts_with(&format!("label: domain module name {name:?} can't be used as-is")), "{refusal}");
        }
    }

    #[test]
    fn accepts_ordinary_names_and_cargo_keys_as_aggregate_names() {
        assert_eq!(reserved_name_refusal("shop", "shop", &["Pizza", "Version", "Default"]), None);
    }
}
